### DISCUSSION TITLE
Running on Windows?

### DISCUSSION THREAD

**Discussion by skybyke** [pashpashpash/vault-ai#44] [Q&A] [answered]
I'd really like to get this running on Windows, but have been struggling getting the right changes made. I've tried adjusting scripts, using Git Bash, but haven't quite been successful. I am going to try and use Docker next, but any help would be super awesome. 

-s

**Comment by dan-dean** [comment]
Through a few modifications, I've gotten it 96% of the way there using git bash and adjusting scripts, same as you. The .exe binary listens on the correct port, but now I'm sorting out issues with API keys and pinecone issues that might just be a me issue. If I finish my changes and get it working fully, I'll try to push it up to my fork of the repo.

**Comment by dan-dean** [comment, marked as answer]
Okay, I've gotten it working but it ended up being a mess and a half. I didn't document all the troubleshooting that I had to do to get it to work with my already messy windows environment, but what matters is that it works. I am mid-finals week so tonight I can't get you what you need, but it is possible. I'll try to push up to my fork in the following week, with a better explanation of what you need to do differently on Windows.
![image](https://user-images.githubusercontent.com/78292223/234187009-9bbfd9d3-3bfa-413b-8f2a-37117ae6fe0e.png)

**Reply by skybyke** [reply]
hey, nice work! i appreciate you digging in and giving it a go. looking forward to getting your output. good luck on those mid-finals!

**Comment by dan-dean** [comment]
https://github.com/dan-dean/vault-ai-windows This should work for you, follow the slightly modified procedure in the README.md. Let me know if you run into any issues

**Reply by skybyke** [reply]
what's the deal with your "enter your openai key here" input field on the front end that appears on first launch, but never again?

**Reply by dan-dean** [reply]
you know, I don't really know. As a whole, did it end up working?

**Reply by skybyke** [reply]
it appeared that it might, yes. which is great. but the field asking for the openai key was incredibly suspicious, so I didn't feel it was worth the risk.

if @pashpashpash can confirm that in the original app there was an "enter your OpenAI key" on the prompt entry page (and if so, explain why) I'll give it a thorough try.

**Reply by dan-dean** [reply]
The original site did have that. You already provide your API key in the environment variable, so if I (or they) were trying to steal that information, prompting for it doesn’t matter. It’s a holdover from his code that allows you to change the API key from when he hosted an instance and allowed people to use his OpenAI credits, but then he ran out so he let people use their own API key on his hosted site. The field is completely optional and only replaces the environmental variable if you fill in the field.
You can find him talking about the optional OpenAI API key field in the comments of his post here:
https://www.reddit.com/r/ChatGPT/comments/12qbrmw/i_built_an_open_source_website_that_allows_you_to/?utm_source=share&utm_medium=ios_app&utm_name=ioscss&utm_content=1&utm_term=1

Once again, I understand the caution but also all the code is there and open to perusal, and even if pashpashpash or I wanted to steal your API key, prompting for it changes nothing since the site already retrieved and uses it from your environment variables.

**Reply by skybyke** [reply]
Understood. I just didn't want to scrounge the code to ensure keys didn't get shipped off anywhere. Yes, I know. A little lazy. Thanks for the reply and acknowledging it was there in the original. I found a thread that might indicate it was, but I just wasn't sure.

That being said, I did get to the prompt screen, eventually, so I think overall it did work. I had some issues with the app reading/finding my OpenAI key and eventually had to set the environment variables directly in the package.json file. 

I'll get it up and running again and take some notes on what I find in case you're curious about the hiccups I had -- and to know whether they are me or you. My environment is a mess as well, honestly, so I have little doubt it's my setup. 

I appreciate the follow up and the work to get here.  Stay tuned!

**Reply by skybyke** [reply]
When i went to do the first query, i'm getting a nil pointer dereference -- related to the retrieve function in pincecone.go. working through that now.

**Reply by DefamationStation** [reply]
> https://github.com/dan-dean/vault-ai-windows This should work for you, follow the slightly modified procedure in the README.md. Let me know if you run into any issues

thanks for this tutorial! any idea why git bash says 
/c/Program Files/nodejs/npm: line 20: node: command not found
/c/Program Files/nodejs/npm: line 23: node: command not found
even though i have nodejs v19 installed?

**Reply by dan-dean** [reply]
Have you tried running node through the command line and also while in bash? It's likely not on your path.

**Reply by DefamationStation** [reply]
I installed the nodj v19 x86 version and it worked, I initially had the x64

**Comment by skybyke** [comment]
I entered a bogus API in the input field, since I thought I had put the real API to be used everywhere else. Now, though it keeps looking for that bogus API key. i cloned the repository again, checked the system variables, opened a new incognito mode browser... i still can't find where it has stored that key. any ideas on how i can clear this ... and why it's not pulling it from somewhere i explicitly set it?

**Reply by pashpashpash** [reply]
@skybyke yeah it's not ideal but the api key is stored in local storage in your browser. If you don't want to update any code you can use a different browser.

**Reply by skybyke** [reply]
thank you. i thought i did that and still got the same, but i blew away all history and data and was able to get where i needed to be. i'm troubleshooting a nil pointer dereference right now.

**Reply by digitalbuddha0101** [reply]
@pashpashpash I have the same problem. Where can I find it in my local browser?

**Comment by dan-dean** [comment]
I've created a pull request with the main project and it should work cross platform, using run-script-os in the package.json file

**Comment by skybyke** [comment]
i'm getting the error below when trying to enter a prompt in the UI. openai suggests that the panic seems to have originated from `github.com/pashpashpash/vault/vault-web-server/postapi.retrieve` function, as indicated by this line:

`github.com/pashpashpash/vault/vault-web-server/postapi.retrieve({0xc000201980, 0x600, 0x63f}, 0xc?, {0xc0002e0000, 0x24})`

Any ideas what might be going wrong here? 

---

"PANIC: runtime error: invalid memory address or nil pointer dereference\ngoroutine 84 [running]:\ngithub.com/codegangsta/negroni.(*Recovery).ServeHTTP.func1()\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/recovery.go:159 +0xc7\npanic({0x12cfd40, 0x18200d0})\n\tC:/Program Files/Go/src/runtime/panic.go:884 +0x213\ngithub.com/pashpashpash/vault/vault-web-server/postapi.retrieve({0xc000201980, 0x600, 0x63f}, 0xc?, {0xc0002e0000, 0x24})\n\tH:/AI/vault-ai-windows-2/vault-ai-windows/vault-web-server/postapi/pinecone.go:131 +0x28d\ngithub.com/pashpashpash/vault/vault-web-server/postapi.QuestionHandler({0x287e8353fe8, 0xc00000a030}, 0xc0003b6210?)\n\tH:/AI/vault-ai-windows-2/vault-ai-windows/vault-web-server/postapi/questions.go:52 +0x4ab\nnet/http.HandlerFunc.ServeHTTP(0xc000292300?, {0x287e8353fe8?, 0xc00000a030?}, 0xc0001a3668?)\n\tC:/Program Files/Go/src/net/http/server.go:2122 +0x2f\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc0000000c0, {0x287e8353fe8, 0xc00000a030}, 0xc000292100)\n\tC:/Users/skyby/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf\ngithub.com/codegangsta/negroni.Wrap.func1({0x287e8353fe8, 0xc00000a030}, 0x12e7f20?, 0xc000322080)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:46 +0x4b\ngithub.com/codegangsta/negroni.HandlerFunc.ServeHTTP(0x10cdc9f?, {0x287e8353fe8?, 0xc00000a030?}, 0xc000524020?, 0xc0001a37d0?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:29 +0x33\ngithub.com/codegangsta/negroni.middleware.ServeHTTP({{0x14355c0?, 0xc000008918?}, 0xc000008960?}, {0x287e8353fe8, 0xc00000a030}, 0xe61297?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:38 +0xb6\ngithub.com/codegangsta/negroni.(*Logger).ServeHTTP(0xc0001fe330, {0x287e8353fe8?, 0xc00000a030}, 0xc000292100, 0xc000322040)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/logger.go:62 +0x8f\ngithub.com/codegangsta/negroni.middleware.ServeHTTP({{0x1434860?, 0xc0001fe330?}, 0xc000008948?}, {0x287e8353fe8, 0xc00000a030}, 0x12e7f20?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:38 +0xb6\ngithub.com/codegangsta/negroni.(*Recovery).ServeHTTP(0x8?, {0x287e8353fe8?, 0xc00000a030?}, 0x0?, 0xc0001a3a50?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/recovery.go:193 +0x86\ngithub.com/codegangsta/negroni.middleware.ServeHTTP({{0x14348a0?, 0xc00007f590?}, 0xc000008930?}, {0x287e8353fe8, 0xc00000a030}, 0xc0000b61b0?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:38 +0xb6\ngithub.com/codegangsta/negroni.(*Negroni).ServeHTTP(0xc0001fe300, {0x1438280?, 0xc0000b61c0}, 0x10cfdb4?)\n\tC:/Users/skyby/go/pkg/mod/github.com/codegangsta/negroni@v1.0.0/negroni.go:96 +0x125\nnet/http.serverHandler.ServeHTTP({0x1436c88?}, {0x1438280, 0xc0000b61c0}, 0xc000292100)\n\tC:/Program Files/Go/src/net/http/server.go:2936 +0x316\nnet/http.(*conn).serve(0xc000294360, {0x1438a58, 0xc0001ff290})\n\tC:/Program Files/Go/src/net/http/server.go:1995 +0x612\ncreated by net/http.(*Server).Serve\n\tC:/Program Files/Go/src/net/http/server.go:3089 +0x5ed\n

**Comment by mgmandahl** [comment]
I wish I saw this thread before I spent two days trying to get this to work. It works now. It took me forever to get it to start up. Finally, working with chatGPT figured out to just bypass the .bat files and load the environment manually, and setting OpenAI and Pinecone manually as well. It's pretty amazing!

**Comment by digitalbuddha0101** [comment]
Thanks for creating this fork. I was trying to install the master on my Windows system and couldn't get it to work. I now have the front-end running but it is returning an error when trying to upload a doc:

Failed Files:
[file name]
Reason: Error getting embeddings: error, status code: 429, message: You exceeded your current quota, please check your plan and billing details.

Is this quota related to The Vault, Pinecone, or something else that I'm missing? I had the thought that perhaps I need to state my The Vault API key somewhere in the package, but I didn't see instructions for that. This is my first pdf and the file size is very small (17kb). Any advice is most welcome.
